File: app/src/main/java/com/centaury/cataloguemovie/ui/favorite/fragment/FavoriteTVShowFragment.java
package com.centaury.cataloguemovie.ui.favorite.fragment;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.centaury.cataloguemovie.R;
import com.centaury.cataloguemovie.data.local.entity.TVShowEntity;
import com.centaury.cataloguemovie.di.Injectable;
import com.centaury.cataloguemovie.ui.favorite.adapter.FavoriteFragmentCallback;
import com.centaury.cataloguemovie.ui.favorite.adapter.FavoriteTVShowAdapter;
import com.centaury.cataloguemovie.ui.favorite.viewmodel.FavoriteTVShowViewModel;
import com.centaury.cataloguemovie.utils.Helper;
import com.facebook.shimmer.ShimmerFrameLayout;

import java.util.concurrent.ExecutionException;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;


/**
 * A simple {@link Fragment} subclass.
 */
public class FavoriteTVShowFragment extends Fragment implements FavoriteFragmentCallback, Injectable {

    @BindView(R.id.rv_favtvshow)
    RecyclerView mRvFavTvshow;
    @BindView(R.id.shimmer_view_container)
    ShimmerFrameLayout mShimmerViewContainer;
    @BindView(R.id.empty_state)
    LinearLayout mEmptyState;
    private Unbinder unbinder;

    @Inject
    ViewModelProvider.Factory factory;

    private FavoriteTVShowViewModel favoriteTVShowViewModel;

    public FavoriteTVShowFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_favorite_tvshow, container, false);

        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getActivity() != null) {
            favoriteTVShowViewModel = obtainViewModel(getActivity());

            FavoriteTVShowAdapter favoriteTVShowAdapter = new FavoriteTVShowAdapter(getActivity(), this);
            favoriteTVShowViewModel.getFavoriteTVShow().observe(getActivity(), favoriteTVShow -> {
                if (favoriteTVShow != null) {
                    mShimmerViewContainer.stopShimmer();
                    mShimmerViewContainer.setVisibility(View.GONE);
                    toggleEmptyMovies(favoriteTVShow.size());
                    favoriteTVShowAdapter.submitList(favoriteTVShow);
                    favoriteTVShowAdapter.notifyDataSetChanged();
                }
            });

            mRvFavTvshow.setLayoutManager(new LinearLayoutManager(getContext()));
            mRvFavTvshow.setHasFixedSize(true);
            mRvFavTvshow.setAdapter(favoriteTVShowAdapter);
            mRvFavTvshow.addItemDecoration(new Helper.TopItemDecoration(55));
        }
    }

    private void showDialogDeleteFavorite(int movieId) {
        TVShowEntity tvShowEntity;
        try {
            tvShowEntity = favoriteTVShowViewModel.getDetailFavTVShow(movieId);

            LayoutInflater inflater = LayoutInflater.from(getContext());
            View view = inflater.inflate(R.layout.item_alert_dialog, null);

            AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
            builder.setView(view);

            TextView title = view.findViewById(R.id.alerttitle);
            title.setText(getString(R.string.txt_title_delete_dialog));

            builder.setCancelable(false)
                    .setPositiveButton(getString(R.string.btn_delete), (dialog, which) -> {
                        favoriteTVShowViewModel.deleteFavoriteTVShow(tvShowEntity);
                        dialog.dismiss();
                        Toast.makeText(getContext(), getString(R.string.txt_remove_movie), Toast.LENGTH_SHORT).show();
                    })
                    .setNegativeButton(getString(R.string.btn_cancel), (dialog, which) -> {
                        dialog.dismiss();
                    });
            AlertDialog alertDialog = builder.create();
            alertDialog.show();

        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }
    }

    @NonNull
    private FavoriteTVShowViewModel obtainViewModel(FragmentActivity activity) {
        return ViewModelProviders.of(activity, factory).get(FavoriteTVShowViewModel.class);
    }

    private void toggleEmptyMovies(int size) {
        if (size > 0) {
            mEmptyState.setVisibility(View.GONE);
            mRvFavTvshow.setVisibility(View.VISIBLE);
        } else {
            mRvFavTvshow.setVisibility(View.GONE);
            mEmptyState.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mShimmerViewContainer.startShimmer();
    }

    @Override
    public void onPause() {
        mShimmerViewContainer.stopShimmer();
        super.onPause();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDeleteItemClick(int movieId) {
        if (getActivity() != null) {
            showDialogDeleteFavorite(movieId);
        }
    }
}

File: app/src/main/java/com/centaury/cataloguemovie/ui/favorite/fragment/FavoriteTVShowFragment.kt
package com.centaury.cataloguemovie.ui.favorite.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.centaury.cataloguemovie.MovieCatalogueApp
import com.centaury.cataloguemovie.R
import com.centaury.cataloguemovie.databinding.FragmentFavoriteTvshowBinding
import com.centaury.cataloguemovie.di.component.DaggerFavoriteTVShowComponent
import com.centaury.cataloguemovie.ui.detail.DetailFavoriteMovieActivity
import com.centaury.cataloguemovie.ui.favorite.adapter.FavoriteFragmentCallback
import com.centaury.cataloguemovie.ui.favorite.adapter.FavoriteTVShowAdapter
import com.centaury.cataloguemovie.ui.favorite.viewmodel.FavoriteTVShowViewModel
import com.centaury.cataloguemovie.utils.CommonUtils
import com.centaury.cataloguemovie.utils.LoaderState
import com.centaury.cataloguemovie.utils.showToast
import com.centaury.domain.tvshow.model.TVShowsEntity
import javax.inject.Inject

/**
 * A simple [Fragment] subclass.
 */
class FavoriteTVShowFragment : Fragment(), FavoriteFragmentCallback {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    private lateinit var favoriteTVShowViewModel: FavoriteTVShowViewModel

    private lateinit var binding: FragmentFavoriteTvshowBinding

    private var tvShowFavoriteData = arrayListOf<TVShowsEntity>()
    private val favoriteTVShowAdapter: FavoriteTVShowAdapter by lazy {
        FavoriteTVShowAdapter(tvShowFavoriteData, this)
    }

    private lateinit var titleDialog: TextView
    private var tvShow: TVShowsEntity? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        binding = FragmentFavoriteTvshowBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initInjector()
        initView(binding)
    }

    private fun initView(binding: FragmentFavoriteTvshowBinding) {
        favoriteTVShowViewModel =
            ViewModelProvider(this, viewModelFactory)[FavoriteTVShowViewModel::class.java]

        with(binding.rvFavTvShow) {
            hasFixedSize()
            layoutManager = LinearLayoutManager(context)
            addItemDecoration(CommonUtils.TopItemDecoration(55))
            adapter = favoriteTVShowAdapter
        }

        initObserver(binding)
    }

    private fun initObserver(binding: FragmentFavoriteTvshowBinding) {
        favoriteTVShowViewModel.state.observe(viewLifecycleOwner, { state ->
            when (state) {
                is LoaderState.ShowLoading -> {
                    binding.shimmerViewContainer.startShimmer()
                    binding.hasFavoriteTVShows = true
                }
                is LoaderState.HideLoading -> {
                    binding.shimmerViewContainer.stopShimmer()
                    binding.hasFavoriteTVShows = false
                }
            }
        })

        favoriteTVShowViewModel.result.observe(viewLifecycleOwner, { result ->
            tvShowFavoriteData.clear()
            tvShowFavoriteData.addAll(result)
            CommonUtils.toggleEmptyState(result.size, binding.emptyState, binding.rvFavTvShow)
            favoriteTVShowAdapter.notifyDataSetChanged()
        })

        favoriteTVShowViewModel.error.observe(viewLifecycleOwner, { error ->
            context?.showToast(error)
        })

        favoriteTVShowViewModel.resultTVShowById.observe(viewLifecycleOwner, { resultTVShowById ->
            tvShow = resultTVShowById
        })

        favoriteTVShowViewModel.errorTVShowById.observe(viewLifecycleOwner, { errorTVShowById ->
            context?.showToast(errorTVShowById)
        })

        favoriteTVShowViewModel.errorDeleteTVShow.observe(viewLifecycleOwner, { errorDeleteTVShow ->
            context?.showToast(errorDeleteTVShow)
        })
    }

    private fun initInjector() {
        DaggerFavoriteTVShowComponent.builder()
            .appComponent((activity?.application as MovieCatalogueApp).appComponent)
            .build()
            .inject(this)
    }

    private fun showDialogDeleteFavorite(tvShowId: Int, position: Int) {
        favoriteTVShowViewModel.getFavoriteTVShowByIdContract(tvShowId)

        val customDialog: AlertDialog.Builder? = context?.let { AlertDialog.Builder(it) }
        val inflater = LayoutInflater.from(context)
        val view = inflater.inflate(R.layout.item_alert_dialog, null)

        titleDialog = view.findViewById(R.id.alert_title)
        titleDialog.text = context?.getString(R.string.txt_title_delete_dialog)

        customDialog?.apply {
            setView(view)
            setCancelable(false)
            setPositiveButton(R.string.btn_delete) { dialog, _ ->
                tvShow?.let { favoriteTVShowViewModel.getDeleteFavoriteTVShowContract(context, it) }
                favoriteTVShowAdapter.notifyItemRemoved(position)
                dialog.dismiss()
                context.showToast(R.string.txt_movie_remove)
            }
            setNegativeButton(R.string.btn_cancel) { dialog, _ ->
                dialog.dismiss()
            }

        }
        val alertDialog = customDialog?.create()
        alertDialog?.show()
    }

    private fun showTransitionImage(movieId: Int, image: ImageView, title: String) {
        val intent = Intent(context, DetailFavoriteMovieActivity::class.java).apply {
            putExtra(DetailFavoriteMovieActivity.DETAIL_EXTRA_FAV_TV_SHOW, movieId)
        }
        activity?.let {
            it.startActivity(
                intent,
                CommonUtils.pairOptionsTransition(it, image, title).toBundle()
            )
        }
    }

    override fun onResume() {
        super.onResume()
        binding.shimmerViewContainer.startShimmer()
    }

    override fun onPause() {
        binding.shimmerViewContainer.stopShimmer()
        super.onPause()
    }

    override fun onDeleteItemClick(movieId: Int, position: Int) {
        showDialogDeleteFavorite(movieId, position)
    }

    override fun onItemClick(movieId: Int, image: ImageView, title: String) {
        showTransitionImage(movieId, image, title)
    }
}